adapters/undertone/undertone.go

```diff
@@ -0,0 +1,214 @@
+}
+
+type BidRequestExt struct {
+	Prebid *openrtb_ext.ExtRequestPrebid `json:"prebid"`
+}
+
+type UndertoneBidderParams struct {
+	Id      int    `json:"id"`
+	Version string `json:"version"`
+}
+
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {
+	bidder := &UndertoneAdapter{
+		endpoint: config.Endpoint,
+	}
+	return bidder, nil
+}
+
+func (a *UndertoneAdapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	err := a.validateBidRequest(request)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	imps, publisherId, errors := a.parseImps(request)
+	if len(imps) == 0 {
+		return nil, errors
+	}
+
+	reqCopy := *request
+	reqCopy.Imp = imps
+
+	a.populateSiteApp(&reqCopy, publisherId, request.Site, request.App)
+	a.populateBidReqExt(&reqCopy)
+
+	requestJSON, err := json.Marshal(reqCopy)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	requestData := &adapters.RequestData{
+		Method: "POST",
+		Uri:    a.endpoint,
+		Body:   requestJSON,
+	}
+
+	return []*adapters.RequestData{requestData}, errors
+}
+
+func (a *UndertoneAdapter) MakeBids(request *openrtb2.BidRequest, requestData *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if responseData.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if responseData.StatusCode == http.StatusBadRequest {
+		err := &errortypes.BadInput{
+			Message: "Unexpected status code: 400. Bad request from publisher. Run with request.debug = 1 for more info.",
+		}
+		return nil, []error{err}
+	}
+
+	if responseData.StatusCode != http.StatusOK {
+		err := &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info.", responseData.StatusCode),
+		}
+		return nil, []error{err}
+	}
+
+	var response openrtb2.BidResponse
+	if err := json.Unmarshal(responseData.Body, &response); err != nil {
+		return nil, []error{err}
+	}
+
+	impIdBidTypeMap := map[string]openrtb_ext.BidType{}
+	for _, imp := range request.Imp {
+		if imp.Banner != nil {
+			impIdBidTypeMap[imp.ID] = openrtb_ext.BidTypeBanner
+		} else if imp.Video != nil {
+			impIdBidTypeMap[imp.ID] = openrtb_ext.BidTypeVideo
+		}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
+	bidResponse.Currency = response.Cur
+	for _, seatBid := range response.SeatBid {
+		for i, bid := range seatBid.Bid {
+			bidType, ok := impIdBidTypeMap[bid.ImpID]
+			if !ok {
+				continue
+			}
+
+			b := &adapters.TypedBid{
+				Bid:     &seatBid.Bid[i],
+				BidType: bidType,
+			}
+			bidResponse.Bids = append(bidResponse.Bids, b)
+		}
+		return bidResponse, nil
+	}
+	return bidResponse, nil
+}
+
+func (a *UndertoneAdapter) populateBidReqExt(bidRequest *openrtb2.BidRequest) {
+	undertoneBidderParams := &UndertoneBidderParams{
+		Id:      4,
+		Version: "1.0.0",
```

Comment:
These should be constants to be declared on the top of this file.
Also we do not need UndertoneBidderParams to be an exported struct since this is not being used anywhere outside of this package.